src/components/Match/renderMatch.js

```diff
@@ -19,11 +17,124 @@ Object.keys(specific).forEach((key) => {
-};
+const getMaxKeyOfObject = (field) =>
+  (field ? Object.keys(field).sort((a, b) => Number(b) - Number(a))[0] : '');
+
+function getLevelFromXp(xp) {
+  for (let i = 0; i < xpLevel.length; i++) {
+    if (xpLevel[i] > xp) {
+      return i;
+    }
+  }
+  return xpLevel.length;
+}
+
+/**
+ * Generates data for c3 charts in a match
+ **/
+function generateGraphData(match) {
+  if (match.players && match.players[0] && match.radiant_gold_adv && match.radiant_xp_adv) {
+    // compute graphs
+    const goldDifference = ['Gold', ...match.radiant_gold_adv];
+    const xpDifference = ['XP', ...match.radiant_xp_adv];
+    const time = ['time', ...match.players[0].times];
+    const data = {
+      difference: [time, xpDifference, goldDifference],
+      gold: [time],
+      xp: [time],
+      lh: [time],
+    };
+    match.players.forEach((p) => {
+      let hero = heroes[p.hero_id] || {};
+      hero = hero.localized_name;
+      if (p.gold_t) {
+        data.gold.push([hero, ...p.gold_t]);
+      }
+      if (p.xp_t) {
+        data.xp.push([hero, ...p.xp_t]);
+      }
+      if (p.lh_t) {
+        data.lh.push([hero, ...p.lh_t]);
+      }
+    });
+    return data;
+  }
+  return {};
+}
+
+/**
+ * Generates position data for a player
+ **/
+function generatePositionData(d, p) {
+  // d, a hash of keys to process
+  // p, a player containing keys with values as position hashes
+  // stores the resulting arrays in the keys of d
+  // 64 is the offset of x and y values
+  // subtracting y from 127 inverts from bottom/left origin to top/left origin
+  const result = {};
+  Object.keys(d).forEach(key => {
+    const t = [];
+    Object.keys(p[key]).forEach(x => {
+      Object.keys(p[key][x]).forEach(y => {
+        t.push({
+          x: Number(x) - 64,
+          y: 127 - (Number(y) - 64),
+          value: p[key][x][y],
+        });
+      });
+    });
+    result[key] = t;
+  });
+  return result;
+}
+
+function generateTeamfights(m) {
+  const computeTfData = (tf) => {
+    const newtf = {};
+    newtf.posData = [];
+    newtf.radiant_gold_delta = 0;
+    newtf.radiant_xp_delta = 0;
+    newtf.radiant_participation = 0;
+    newtf.radiant_deaths = 0;
+    newtf.dire_participation = 0;
+    newtf.dire_deaths = 0;
+    newtf.players = m.players.map(p => {
+      const tfplayer = tf.players[p.player_slot % (128 - 5)];
+      // compute team gold/xp deltas
+      if (isRadiant(p.player_slot)) {
+        newtf.radiant_gold_delta += tfplayer.gold_delta;
+        newtf.radiant_xp_delta += tfplayer.xp_delta;
+        newtf.radiant_participation += tfplayer.participate ? 1 : 0;
+        newtf.radiant_deaths += tfplayer.deaths ? 1 : 0;
+      } else {
+        newtf.radiant_gold_delta -= tfplayer.gold_delta;
+        newtf.radiant_xp_delta -= tfplayer.xp_delta;
+        newtf.dire_participation += tfplayer.participate ? 1 : 0;
+        newtf.dire_deaths += tfplayer.deaths ? 1 : 0;
+      }
+      return {
+        ...p,
+        ...tfplayer,
```

Comment:
p is probably player or something, but in order for me to figure out what it is, I have to scroll all the way up the file.